### PR TITLE
Enable fast softmax mode in FusedSDPA

### PR DESCRIPTION
**Pull request by wszczurekhabana**
Support for setting fast softmax mode in FusedSDPA operator. This is a tradeoff: performance vs accuracy.

Data on performance:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/wszczure/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/wszczure/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

Ratio | Max input tokens | Max new tokens | Batch size | Throughput without fast softmax   [tokens/s] | Throughput with fast softmax   [tokens/s] | Improvement %
-- | -- | -- | -- | -- | -- | --
97% | 31744 | 1042 | 12 | 139.08 | 147.97 | 6.4%
75% | 24576 | 8192 | 16 | 431.09 | 437.95 | 1.6%
50% | 16384 | 16384 | 24 | 653.39 | 656.38 | 0.5%

</body>

</html>

Data on accuracy (using mlperf test from: https://gerrit.habana-labs.com/plugins/gitiles/mlperf_inference/+/refs/heads/master_next/code/llama/llama_greedy.py
and https://gerrit.habana-labs.com/plugins/gitiles/mlperf_inference/+/refs/heads/master_next/code/llama/evaluation.py):

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/wszczure/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/wszczure/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

  | rouge1 | rouge2 | rougeL | rougeLsum | accuracy
-- | -- | -- | -- | -- | --
without fast softmax | 44.4279 | 22.0536 | 28.6362 | 42.0044 | 99.99
with fast softmax | 44.4065 | 22.0229 | 28.6156 | 41.9858 | 99.94

</body>

</html>

### CODE REVIEW

**Review by dvarshney-habana — APPROVED** [review]

File: optimum/habana/transformers/models/llama/modeling_llama.py
Comment (puneeshkhanna) [inline optimum/habana/transformers/models/llama/modeling_llama.py:766]: Please rebase and add argument to self._gradient_checkpointing_func too

**Comment by dudilester** [comment]
This ModuleFusedSDPA forward API change will require changes in the HQT patched module for quantization. which means it will break the nightly testing once merged. Im not sure we support regular-softmax for 8bit, we need to consider the appropriate behavior when user requests both quantization and regular-softmax,  should we ignore the quantization or the softmax? or assert on that configuration.

File: optimum/habana/transformers/models/llama/modeling_llama.py
Comment (wszczurekhabana) [reply optimum/habana/transformers/models/llama/modeling_llama.py:771]: Done in latest commits.

**Comment by wszczurekhabana** [comment]
Discussed offline. Relevant change for quantization toolkit: https://gerrit.habana-labs.com/#/c/411008/ pushed by @dudilester is in review.

**Comment by wszczurekhabana** [comment]
Change in https://gerrit.habana-labs.com/#/c/411008/ is merged. @dvarshney-habana @puneeshkhanna @dudilester I think we can merge this PR now.

**Comment by dudilester** [comment]
Change in https://gerrit.habana-labs.com/#/c/411008/ did not pass promotion yet, we need to wait till it will pass before we merge this PR.

**Comment by dudilester** [comment]
FYI, commit https://gerrit.habana-labs.com/#/c/411008/ was promoted since my previous comment, and is included in builds since CD 1.16.0-328 release build

**Comment by wszczurekhabana** [comment]
upstreamed in: https://github.com/huggingface/optimum-habana/pull/972

**Pull request merged**
